python/huggingfaceserver/huggingfaceserver/vllm_completions.py

```diff
@@ -0,0 +1,452 @@
+from kserve.protocol.rest.openai.types.openapi import (
+    CreateCompletionRequest,
+    ErrorResponse,
+    Error,
+    Logprobs,
+)
+
+
+def to_sampling_params(request: CreateCompletionRequest):
+    echo_without_generation = request.echo and request.max_tokens == 0
+
+    logits_processors = None
+    if request.logit_bias:
+
+        def logit_bias_logits_processor(
+            token_ids: List[int], logits: torch.Tensor
+        ) -> torch.Tensor:
+            for token_id, bias in request.logit_bias.items():
+                # Clamp the bias between -100 and 100 per OpenAI API spec
+                bias = min(100, max(-100, bias))
+                logits[int(token_id)] += bias
+            return logits
+
+        logits_processors = [logit_bias_logits_processor]
+
+    return SamplingParams(
+        n=request.n,
+        best_of=request.best_of,
+        presence_penalty=request.presence_penalty,
+        frequency_penalty=request.frequency_penalty,
+        temperature=request.temperature,
+        top_p=request.top_p,
+        seed=request.seed,
+        stop=request.stop,
+        logprobs=request.logprobs,
+        max_tokens=request.max_tokens if not echo_without_generation else 1,
+        logits_processors=logits_processors,
+        prompt_logprobs=request.logprobs if request.echo else None,
+    )
+
+
+class OpenAIServingCompletion:
+
+    def __init__(self, engine: AsyncLLMEngine, served_model: str):
+        self.engine = engine
+        self.served_model = served_model
+
+        self.max_model_len = 0
+        self.tokenizer = None
+
+        try:
+            event_loop = asyncio.get_running_loop()
+        except RuntimeError:
+            event_loop = None
+
+        if (
+            event_loop is not None and event_loop.is_running()
+        ):  # If the current is instanced by Ray Serve, there is already a running event loop
+            event_loop.create_task(self._post_init())
+        else:  # When using single vLLM without engine_use_ray
+            loop = asyncio.new_event_loop()
+            loop.run_until_complete(self._post_init())
+            loop.close()
+
+    async def create_completion(self, request: CreateCompletionRequest):
+        """Completion API similar to OpenAI's API.
+
+        See https://platform.openai.com/docs/api-reference/completions/create
+        for the API specification. This API mimics the OpenAI Completion API.
+
+        NOTE: Currently we do not support the following feature:
+            - suffix (the language models we currently support do not support
+            suffix)
+        """
+        error_check_ret = await self._check_model(request)
+        if error_check_ret is not None:
+            return error_check_ret
+
+        # Return error for unsupported features.
+        if request.suffix is not None:
+            return self.create_error_response("suffix is not currently supported")
+
+        model_name = request.model
+        request_id = f"cmpl-{random_uuid()}"
```

Comment:
Request id may need to be fetched from the header `x-request-id` if user sets from the client side, need to think about how we can pass the headers like KServe model does.

Reply:
Will need to pass through here, right?
https://github.com/kserve/kserve/blob/master/python/kserve/kserve/protocol/rest/openai/openai_model.py#L62

Reply:
Updated with the changes